### PR TITLE
Add more bindings to OpenSSL, to be used by the Python _ssl module.

### PR DESCRIPTION
**Pull request by amauryfa**
I am trying to implement the Python `_ssl`  module using cffi bindings:
PyPy has an RPython implementation of this module, so the main executable is statically linked with some version of OpenSSL, which can collide with other extension modules.

These bindings are enough to implement most of the SSL features, and I could write a `_ssl.py` module (not included here) which passes many tests in test_ssl.py.
Not implemented: the parts related to "certificate store": `ctx.get_ca_certs()`, `ctx.cert_store_stats()`.

### CODE REVIEW

**Comment by alex** [comment]
Jenkins, ok to test.

**Comment by alex** [comment]
@amauryfa can you merge `master` into this branch? I believe there's some conflicts.

**Comment by reaperhulk** [comment]
This LGTM. Once it merges I'll check it against my 1.1.0-pre4 builds, but on first glance it looks like there shouldn't be any major issues.

File: src/_cffi_src/openssl/ssl.py [hunk 1]
```diff
@@ -435,6 +447,19 @@
                                   size_t);
 size_t SSL_get_client_random(const SSL *, unsigned char *, size_t);
 size_t SSL_get_server_random(const SSL *, unsigned char *, size_t);
+
+long SSL_CTX_sess_number(SSL_CTX *ctx);
```
Comment (reaperhulk) [inline src/_cffi_src/openssl/ssl.py]: These all need to have their arg names removed.

File: src/_cffi_src/openssl/ssl.py [hunk 2]
```diff
@@ -356,6 +366,8 @@
 void SSL_CTX_set_tlsext_servername_callback(
     SSL_CTX *,
     int (*)(const SSL *, int *, void *));
+void SSL_CTX_set_tlsext_servername_arg(
+    SSL_CTX *, void*);
```
Comment (reaperhulk) [inline src/_cffi_src/openssl/ssl.py]: Should be a space between `void` and `*`

File: src/_cffi_src/openssl/ssl.py [hunk 3]
```diff
@@ -545,6 +570,8 @@
 void (*SSL_CTX_set_tlsext_servername_callback)(
     SSL_CTX *,
     int (*)(const SSL *, int *, void *)) = NULL;
+void (*SSL_CTX_set_tlsext_servername_arg)(
+    SSL_CTX *, void*) = NULL;
```
Comment (reaperhulk) [inline src/_cffi_src/openssl/ssl.py]: Space between `void` and `*` as well

**Comment by reaperhulk** [comment]
Just a few small style nits. Once those are resolved and tests pass I'll merge. Thanks @amauryfa!

File: src/_cffi_src/openssl/ssl.py (hunk 1 above)
Comment (amauryfa) [inline src/_cffi_src/openssl/ssl.py]: Done

File: src/_cffi_src/openssl/ssl.py (hunk 2 above)
Comment (amauryfa) [inline src/_cffi_src/openssl/ssl.py]: Done

File: src/_cffi_src/openssl/ssl.py (hunk 3 above)
Comment (amauryfa) [inline src/_cffi_src/openssl/ssl.py]: Done

File: src/_cffi_src/openssl/bio.py [hunk 4]
```diff
@@ -158,6 +158,7 @@
 long BIO_set_write_buffer_size(BIO *, long);
 long BIO_set_buffer_size(BIO *, long);
 long BIO_set_buffer_read_data(BIO *, void *, long);
+long BIO_set_nbio(BIO *b, long n);
```
Comment (alex) [inline src/_cffi_src/openssl/bio.py]: Param name should be removed here.

**Comment by alex** [comment]
Thanks @amauryfa !

File: src/_cffi_src/openssl/bio.py (hunk 4 above)
Comment (amauryfa) [inline src/_cffi_src/openssl/bio.py]: Done

**Pull request merged**